subsys/bluetooth/host/hci_core.h

```diff
@@ -479,4 +479,4 @@ void bt_hci_le_df_cte_req_failed(struct net_buf *buf);
 /** First thread that called bt_recv. NULL if not yet called.
  *  Updated non-atomically; may be used only to compare to current thread id.
  */
-extern k_tid_t bt_recv_thread_id;
+extern k_tid_t hci_acl_thread_id;
```

Comment:
This variable name must start with "`bt_`". It's the namespace for this file.

I'm not too happy about the new name since `hci_acl` is an hidden internal / static function. But I will approve `bt_hci_acl_thread_id` if others like it. I think `bt_acl_cb_thread` is better. This variable has always been leaking implementation and would benefit from an abstraction. If anybody has any idea, speak up.

Reply:
@alwa-nordic we don't generally enforce names-spacing for stack-internal symbols, but we don't forbid it either. In fact, it's especially encouraged for internal symbols that are exported between multiple c-files. So I support this suggestion.

Reply:
> Is this value intended for other uses than ATT?

No. This variable is a hack. It has no intended use. We would like to get rid of it if we can.

> Does the issue it solves exist for other fixed L2CAP channels? Does it exist for dynamic L2CAP channels?

Maybe. The GATT API is affected because (1.) it promises to be async, and (2.) it ends up in an allocation call with timeout K_NO_WAIT. If (1.) wasn't true, the deadlock would be the API users fault for calling blocking API from BT RX context.

> Is there any other thread besides the "hci_acl" that calls bt_att_recv?

Possibly; the system work queue. This call path exist:

```
l2cap_rx_process
l2cap_chan_le_recv
l2cap_chan_le_recv_seg
l2cap_chan_le_recv_sdu
bt_att_recv
```

> Is it possible, and does it make sense to move this value to att.c and rename it to something like bt_att_recv_thread_id? Or do we want to keep it more generic?

I assume you are thinking of moving the capture of `k_current_get`? I looked a bit around and I think it can go into `bt_att_accept`. `bt_att_accept` initializes a connection's `struct bt_att`, so it must have been called before the application can make any GATT-request functions, right?

Reply:
> I assume you are thinking of moving the capture of `k_current_get`? I looked a bit around and I think it can go into `bt_att_accept`. `bt_att_accept` initializes a connection's `struct bt_att`, so it must have been called before the application can make any GATT-request functions, right?

Correct. It looks like the call to `k_current_get` could/should be moved to some ATT function instead.

Reply:
Ok. We can move the variable to `att.c` and capture the thread id in `bt_att_accept`. The most meaningful variable I have in mind is the following:

```c
/** The thread ATT response handlers likely run on.
 *
 *  Blocking this thread while waiting for an ATT request to resolve can cause a
 *  deadlock.
 * 
 *  This can happen if the application queues ATT requests in the context of a
 *  callback from the Bluetooth stack. This is because queuing an ATT request
 *  will block until a request-resource is available, and the callbacks run on
 *  the same thread as the ATT response handler that frees request-resources.
 *
 *  The intended use of this value is to detect the above situation.
 */
static k_tid_t att_handle_rsp_thread;
```

@LingaoM Do you think this will function correctly?

Reply:
@LingaoM ping